File: Part3/shared/EllipticCurve/EllipticCurve.cpp
#include <Arduino.h>
#include "EllipticCurve.h"
#include "Math.h"

// Function that performs point adding
Point EllipticCurve::addPoint(const Point& point1, const Point& point2)
{
    if (point1.x == point2.x && point1.y == point2.y) // the same point
        return doublingPoint(point1);  
    if (point1.x == point2.x) // next point will be infinity point
        return InfPoint;    
    if (point1 == InfPoint) // return the second point
        return point2;
    if (point2 == InfPoint) // return the second point
        return point1;
    Point R;
    int32_t numerator = point1.y - point2.y;
    int32_t denominator = point1.x - point2.x;

    if (denominator < 0){
        numerator *= -1;
        denominator *= -1;
    }

    int32_t s = module(numerator, p) * modularInverse(denominator, p) % p;

    R.x = s * s - (point1.x + point2.x);
    R.x = module(R.x, p);
    R.y = s * (point1.x - R.x) - point1.y;
    R.y = module(R.y, p);
    return R;
}

// Function that performs point doubling
Point EllipticCurve::doublingPoint(const Point& point){
    if (point == InfPoint) // InfPoint+InfPoint = InfPoint
        return InfPoint;
    if (point.y == 0) // InfPoint because this is a vertical line
        return InfPoint;
    Point R;
    int32_t numerator = 3 * point.x * point.x + a;
    int32_t denominator = 2 * point.y;
    int32_t s = module(numerator, p) * modularInverse(denominator, p) % p;

    R.x = s * s - 2 * point.x;
    R.x = module(R.x, p);
    R.y = s * (point.x - R.x) - point.y;
    R.y = module(R.y, p);
    return R;
}

// Function that implements montgomery algorithm
Point EllipticCurve::scalarMultiplication(Point P, uint32_t k){
    Point R0 = P;
    byte l = key_length(k);
    for(int8_t j = l-2; j >= 0; j--)
    {
        R0 = doublingPoint(R0);
        if(bitRead(k, j)){ // if k[j] == 1
            R0 = addPoint(R0, P);
        }
        delay(5);
    }
    return R0;
}

// Function to calculate the point on the elliptic curve given the private key
Point EllipticCurve::EllipticCurveCalcPoint(Point P, uint32_t PrivKey){
    return scalarMultiplication(P, PrivKey); // use the scalar multiplication algorithm
}

// this function calculates the length in bits of the key
byte EllipticCurve::key_length(uint32_t k)
{
    for (uint8_t i=31; i>=0;i--)
        if(bitRead(k, i))
            return i + 1;
    return 0;
}

// simple curve order algorithm
uint32_t EllipticCurve::calcOrder(Point G)
{
    Point P = G;
    uint32_t i = 2;
    for (; !(P == InfPoint); i++){ P = addPoint(G, P); }
    return i-1;
}


bool EllipticCurve::isOnCurve(Point G)
{
    return (G.y * G.y) % p == (G.x * G.x * G.x + a * G.x + b) % p;
}

File: old_Part3/Attack/elliptic_curve.py

# class to present a point
class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __eq__(self, other):
        return self.x == other.x and self.y == other.y

    def __str__(self):
        if self.x == -1 and self.y == -1:
            return "Point at infinity"
        return f"({self.x}, {self.y})"
    def __hash__(self):
        return hash((self.x, self.y))

# class to present Elliptic Curve
class EllipticCurve:
    def __init__(self, a, b, p):
        self.p = p
        self.b = b
        self.a = a
        self.InfPoint = Point(-1, -1)

    def addPoint(self, point1, point2):
        if point1.x == point2.x and point1.y == point2.y:
            return self.doublingPoint(point1)
        if point1.x == point2.x: # vertical
            return self.InfPoint
        if point1 == self.InfPoint: 
            return point2
        if point2 == self.InfPoint:
            return point1

        numerator = point1.y - point2.y
        denominator = point1.x - point2.x
        if denominator < 0:
            numerator *= -1
            denominator *= -1
        s = self.module(numerator, self.p) * self.modularInverse(denominator, self.p) % self.p

        R_x = s * s - (point1.x + point2.x)
        R_x = self.module(R_x, self.p)
        R_y = s * (point1.x - R_x) - point1.y
        R_y = self.module(R_y, self.p)
        return Point(R_x, R_y)

    def doublingPoint(self, point):
        if point == self.InfPoint:
            return self.InfPoint
        if point.y == 0:
            return self.InfPoint

        numerator = 3 * point.x * point.x + self.a
        denominator = 2 * point.y
        s = self.module(numerator, self.p) * self.modularInverse(denominator, self.p) % self.p

        R_x = s * s - 2 * point.x
        R_x = self.module(R_x, self.p)
        R_y = s * (point.x - R_x) - point.y
        R_y = self.module(R_y, self.p)
        return Point(R_x, R_y)

    def scalarMultiplication(self, P, k):
        R0 = P
        l = self.key_length(k)
        for j in range(l - 2, -1, -1):
            R0 = self.doublingPoint(R0)
            if (k >> j) & 1:
                R0 = self.addPoint(R0, P)
        return R0

    def EllipticCurveCalcPoint(self, P, PrivKey):
        return self.scalarMultiplication(P, PrivKey)

    def gcdExtended(self, a, b):
        if a == 0:
            return b, 0, 1
        gcd, x1, y1 = self.gcdExtended(b % a, a)
        x = y1 - (b // a) * x1
        y = x1
        return gcd, x, y

    def modularInverse(self, a, p):
        gcd, x, y = self.gcdExtended(a, p)
        if gcd != 1:
            return -1  # Modular inverse doesn't exist
        else:
            return (x % p + p) % p

    def key_length(self, k):
        for i in range(31, -1, -1):
            if (k >> i) & 1: # reach the first non-zero bit
                return i + 1
        return 0

    def module(self, a, b):
        r = a % b
        if r < 0: 
            r += b
        return r

# Example usage
if __name__ == "__main__":
    curve = EllipticCurve(a=2, b=3, p=97)
    P = Point(3, 6)
    R0 = P
    R0 = curve.addPoint(R0, P)
    R0 = curve.addPoint(R0, P)
    R0 = curve.addPoint(R0, P)
    R0 = curve.addPoint(R0, P)
    print(R0)
